compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp

```diff
@@ -0,0 +1,240 @@
+{
+  auto x = loco::must_cast<luci::CircleNode *>(add->x());
+  auto y = loco::must_cast<luci::CircleNode *>(add->y());
+
+  luci::CircleMul *pred = nullptr;
+  luci::CircleConst *constant = nullptr;
+
+  if (add->fusedActivationFunction() != luci::FusedActFunc::RELU)
+    return false;
+
+  if (x->opcode() == luci::CircleOpcode::CIRCLECONST && y->opcode() == luci::CircleOpcode::MUL)
+  {
+    pred = loco::must_cast<luci::CircleMul *>(y);
+    constant = loco::must_cast<luci::CircleConst *>(x);
+  }
+  else if (x->opcode() == luci::CircleOpcode::MUL && y->opcode() == luci::CircleOpcode::CIRCLECONST)
+  {
+    pred = loco::must_cast<luci::CircleMul *>(x);
+    constant = loco::must_cast<luci::CircleConst *>(y);
+  }
+  else
+  {
+    return false;
+  }
+
+  if (constant->rank() != 1)
+    return false;
+
+  auto channel_dim = constant->dim(0);
+  // Assumption: Layout is channel-last
+  if (!(channel_dim == add->dim(add->rank() - 1)))
+    return false;
+
+  mul = pred;
+  beta = constant;
+  return true;
+}
+
+} // namespace
+
+namespace luci
+{
+
+/**
+ *  Swap MUL/ADD if they are from batch normalization
+ *
+ *  BEFORE
+ *           [Mul]  gamma
+ *             |
+ *        [Add + Relu]  beta
+ *
+ *  AFTER
+ *           [Add]  beta/gamma
+ *             |
+ *           [Mul]  gamma
+ *             |
+ *           [Relu]
+ */
+bool swap_mul_add(luci::CircleAdd *add, std::vector<luci::CircleMul *> &mul_list,
+                  std::vector<luci::CircleAdd *> &add_list)
+{
+  luci::CircleNode *pred_node = nullptr;
+  luci::CircleMul *mul = nullptr;
+  luci::CircleConst *beta = nullptr;
+  luci::CircleConst *gamma = nullptr;
+
+  if (!is_batchnorm_add(add, mul, beta))
+    return false;
+
+  if (loco::succs(mul).size() != 1)
+    return false;
+
+  if (!is_batchnorm_mul(mul, pred_node, gamma))
+    return false;
+
+  if (beta->dtype() != loco::DataType::FLOAT32 || gamma->dtype() != loco::DataType::FLOAT32)
+    throw std::runtime_error("FusePreActivationBatchNormPass only supports Float32 model");
+
+  if (!is_non_negative(gamma))
+    return false;
+
+  // Insert Relu at the bottom
+  auto relu = add->graph()->nodes()->create<luci::CircleRelu>();
+  relu->features(mul);
+  loco::replace(add).with(relu);
+
+  // Replace beta <- beta / gamma
+  if (add->x() == beta)
+  {
+    add->y(pred_node);
+  }
+  else
+  {
+    add->x(pred_node);
+  }
+  add->fusedActivationFunction(luci::FusedActFunc::NONE);
+  uint32_t size = beta->size<loco::DataType::FLOAT32>();
+  for (uint32_t i = 0; i < size; ++i)
+  {
+    auto b = beta->at<loco::DataType::FLOAT32>(i);
+    auto g = gamma->at<loco::DataType::FLOAT32>(i);
+    if (b == g)
+    {
+      beta->at<loco::DataType::FLOAT32>(i) = 1;
+    }
+    else
+    {
+      // If g is 0, we use a small value (empirically determined)
+      if (g == 0)
+        g = 1e-10;
+      beta->at<loco::DataType::FLOAT32>(i) = b / g;
+    }
```

Comment:
```suggestion
      // If g is 0, we use a small value (empirically determined)
      if (g == 0)
        g = 1e-10;
      beta->at<loco::DataType::FLOAT32>(i) = b / g;
    
```
Isn't it possible?

Reply:
If b ==0 and g == 0, beta will be 1 in the previous code and 0 in the suggested code.

Actually, this is a corner case that may not happen in real models. I made it to 1 for simple testing (https://github.com/Samsung/ONE/pull/4769/files#diff-ce6d441c9cf90c4ae46084e6c0b00d6d0031c90b455ca36afa662e7e335156b4R179).